kernel: events: introduce safe variants of k_event_wait()

cfriedt opened the pull request:
This change introduces two companion syscalls that allow us to wait for events in a safe way; `k_event_wait_safe()` and `k_event_wait_all_safe()`. These syscalls will remove any (`k_event_wait_safe()`) or all (`k_event_wait_all_safe()`) of the specified events from the event object upon success. 

With that, callers will not receive phantom events. Each time an event is ready, callers can be certain that it was an actual event.

Similarly, there is no explicit need to reset or clear events manually, since the received events are automatically cleared by `k_event_wait_safe()` and `k_event_wait_all_safe()`.

> [!NOTE]  
> Losing event information is equally probably for level-triggered and edge-triggered events. Hence the choice of the `_safe` suffix.

Fixes #89625

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/89624/docs/kernel/services/synchronization/events.html)

Testing Done:
```shell
west build -p auto -b qemu_riscv64 -t run tests/kernel/events/event_api
...
*** Booting Zephyr OS build v4.1.0-3746-g43bdaf9ef17b ***
Running TESTSUITE events_api
===================================================================
START - test_event_deliver
 PASS - test_event_deliver in 0.001 seconds
===================================================================
START - test_event_receive
 PASS - test_event_receive in 0.447 seconds
===================================================================
START - test_k_event_init
 PASS - test_k_event_init in 0.001 seconds
===================================================================
START - test_k_event_wait_all_safe
 PASS - test_k_event_wait_all_safe in 0.001 seconds
===================================================================
START - test_k_event_wait_safe
 PASS - test_k_event_wait_safe in 0.001 seconds
===================================================================
TESTSUITE events_api succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [events_api]: pass = 5, fail = 0, skip = 0, total = 5 duration = 0.451 seconds
 - PASS - [events_api.test_event_deliver] duration = 0.001 seconds
 - PASS - [events_api.test_event_receive] duration = 0.447 seconds
 - PASS - [events_api.test_k_event_init] duration = 0.001 seconds
 - PASS - [events_api.test_k_event_wait_all_safe] duration = 0.001 seconds
 - PASS - [events_api.test_k_event_wait_safe] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

peter-mitsis commented on kernel/events.c:
```diff
@@ -293,16 +305,23 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
-	if (z_pend_curr(&event->lock, key, &event->wait_q, timeout) == 0) {
-		/* Retrieve the set of events that woke the thread */
-		rv = thread->events;
+	for (k_timepoint_t end = sys_timepoint_calc(timeout); !K_TIMEOUT_EQ(timeout, K_NO_WAIT);
+	     timeout = sys_timepoint_timeout(end)) {
+
```
I am assuming that it is meant for some sort or retry logic, which experience has taught me is tricky. Upon the 2nd pass through this loop the event object's spinlock is going to be unlocked, which will violate an assumption of z_pend_curr().

cfriedt replied:
Yeah, I think there is a minor race condition there. Is there any version of pend that retains the lock?

peter-mitsis replied:
No. But one thing I am trying to figure out is why have the retry condition at all?

cfriedt replied:
It's probably not needed anymore. Originally, I thought it might fix something unusual I saw, but that was fixed with a modification to the post codepath

cfriedt replied:
Fixed!

peter-mitsis reviewed (commented):
I will need to go through this a few times more, but so far I have spotted one issue.

TaiJuWu commented on kernel/events.c:
```diff
@@ -166,7 +176,6 @@ static uint32_t k_event_post_internal(struct k_event *event, uint32_t events,
 		struct k_thread *next;
 		do {
 			arch_thread_return_value_set(thread, 0);
-			thread->events = events;
```
Sorry, could you help me understand why we need to remove this line?

cfriedt replied:
Of course - actually it was mainly just moved from `k_event_post_internal()` to `event_walk_op()`, but the main difference is that the actual matched events are stored back in the thread, vs the events that were posted).

TaiJuWu replied:
Got the point, thanks for your explanation.

Commit:
kernel: events: introduce internal K_EVENT_WAIT_REMOVE flag

Introduce the internal K_EVENT_WAIT_REMOVE flag, which instructs
k_event_wait_internal() to remove matched events (all or enay)
from the set of events stored in the event object.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
kernel: events: introduce safe variants of k_event_wait()

Previously, the k_event_wait() and k_event_wait_all() APIs had a
shortcoming, in that they received events could never be cleared in a
safe way.

The only way to clear individual events was with k_event_clear(),
but that is inherently racey, and could result in lost event
information even when using a single thread and an interrupt.

The alternative, setting the "reset" parameter of k_event_wait() or
k_event_wait_all(), was also unsafe since it would clear all events,
leading to event information loss.

Without a safe way to clear events, the application would naturally
experience "phantom" events. I.e. events that were not cleared from
previous k_event_wait() operations and actually never occurred.

Note: since the events are stored as a bitmask and are not queued, it
is not possible to know if events have been issues multiple times
prior to being serviced.

Event information loss, as well as phantom events, could be considered
somewhat unsafe, regardless of whether events were edge triggered
or level triggered.

This change introduces two companion syscalls that allow us to wait for
events in a safe way; k_event_wait_safe() and k_event_wait_all_safe().

These two new system calls use the internal K_EVENT_WAIT_REMOVE flag
to indicate to k_event_wait_internal(), that it should atomically
remove the matching, received events from the set of events contained
in the event object.

With that, callers will not receive phantom events. Each time an
event is ready, callers can be certain that it was an actual event.

Similarly, there is no explicit need to reset or clear events manually,
since the received events are automatically cleared by
k_event_wait_safe() and k_event_wait_all_safe().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: kernel: events_api: test safe versions of k_event_wait()

Add tests for the newly added k_event_wait_safe() and
k_event_wait_all_safe().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: kernel: include k_event_wait_safe() and k_event_wait_all_safe()

Include documentation for k_event_wait_safe() and
k_event_wait_all_safe().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented on doc/kernel/services/synchronization/events.rst:
```diff
@@ -152,13 +152,20 @@ before continuing.
         ...
     }
 
+Both :c:func:`k_event_wait` and :c:func:`k_event_wait_all` clear the receieved
```
Moved the doc change to a separate commit that simply highlights "(without removal)" and "(without removal)" variants of `k_event_wait()`.

(fixed)

jonas-5ae29c replied:
The same typo persists in `k_event_wait_safe` and `k_event_wait_all_safe` docs.

cfriedt replied:
Fixed!

cfriedt commented:
* minor adjustment to language used in `.rst`

cfriedt commented:
* used [recommendation](https://sonarcloud.io/project/issues?id=zephyrproject-rtos_zephyr&pullRequest=89624&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) from sonarcubecloud

cfriedt commented:
* removed `reset` parameter in `z_vrfy_` wrapper

cfriedt commented:
* added `_safe` suffix in `z_vrfy_` wrapper

henrikbrixandersen commented:
Would `_edge` be a better suffix?

cfriedt commented:
> Would `_edge` be a better suffix? 

I'm open to `_edge` too. I chose `_safe` here simply because it's still possible for level-triggered events (i.e. the non-safe versions) to lose event information with the previous API.

Would be great to get additional feedback from @peter-mitsis and @andyross before the next revision.

cfriedt commented:
CI seems to be stuck on 
```
Requested runner group: zephyr-runner-v2-linux-x64-4xlarge
Job defined at: zephyrproject-rtos/zephyr/.github/workflows/twister.yaml@refs/pull/89624/merge
Waiting for a runner to pick up this job...
Job is waiting for a runner from 'zrv2-linux-x64-4xlarge-hzr' to come online.
```

peter-mitsis commented on kernel/events.c:
```diff
@@ -294,15 +304,17 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
 	if (z_pend_curr(&event->lock, key, &event->wait_q, timeout) == 0) {
-		/* Retrieve the set of events that woke the thread */
 		rv = thread->events;
+		if (options & K_EVENT_WAIT_REMOVE) {
+			/* Remove events that have been received */
+			event->events &= ~rv;
+		}
```
I think this has race conditions.

Scenario 1. We do not know when this thread is going to get scheduled next. Clearing the event object's events may clear events that have occurred between when the events were delivered to the event object and when the thread gets to execute.

Scenario 2. (Really a special case of the previous) The event object is no longer protected by the spinlock. A context switch could occur during the AND'ing (after the events were loaded into a register, but before they were stored to memory). New events may be lost once this thread is scheduled again.

cfriedt replied:
Would be good to figure out a way to tighten it down.

ckhardin replied:
clear the events in the data walk - here is a the code that is pretty close to this patch

```c
static int event_walk_op(struct k_thread *thread, void *data)
{
        uint32_t match;
        unsigned int wait_condition;
        struct event_walk_data *event_data = data;

        wait_condition = thread->event_options & K_EVENT_WAIT_MASK;

        match = are_wait_conditions_met(thread->wevents, event_data->events,
                                        wait_condition);
        if (match != 0) {
                /*
                 * Events create a list of threads to wake up. We do
                 * not want z_thread_timeout to wake these threads; they
                 * will be woken up by k_event_post_internal once they
                 * have been processed.
                 */
                thread->no_wake_on_timeout = true;
                z_abort_timeout(&thread->base.timeout);

                /*
                 * The wait conditions have been satisfied. So, set the
                 * received events and then add this thread to the list
                 * of threads to unpend.
                 *
                 * NOTE: thread event options can consume an event
                 */
                thread->revents = match;
                if (thread->event_options & K_EVENT_OPTION_CLEAR) {
                        event_data->clear_events |= match;
                }
                thread->next_event_link = event_data->head;
                event_data->head = thread;
        }

        return 0;
}
```

ckhardin commented:
we need to get this api to be a bit more resilient with regards to phantom events - so, it ok to get the events and then we can use a "set_masked" or "clear" in the thread to get the current events but this is a diff that is similar to this but closes the races on the clear of the events to only occur when an event is consumed into a thread via a wait call or a "wake walk" on the thread list.

```diff
diff --git a/doc/kernel/services/synchronization/events.rst b/doc/kernel/services/synchronization/events.rst
index cbb86894723..c7dc43387ae 100644
--- a/doc/kernel/services/synchronization/events.rst
+++ b/doc/kernel/services/synchronization/events.rst
@@ -108,8 +108,8 @@ the event object.
         ...
     }
 
-Waiting for Events
-==================
+Waiting for Events (without removal)
+====================================
 
 Threads wait for events by calling :c:func:`k_event_wait`.
 
@@ -152,6 +152,53 @@ before continuing.
         ...
     }
 
+Waiting for Events (with removal)
+=================================
+
+Threads wait for events (with atomic removal upon receipt) by calling
+:c:func:`k_event_wait_safe`.
+
+The following code builds on the example above, and waits up to 50 milliseconds
+for any of the specified events to be posted.  A warning is issued if none
+of the events are posted in time.
+
+If events are received on time, then they will not be present in the event
+object until the next time that the events are set or posted.
+
+.. code-block:: c
+    void consumer_thread(void)
+    {
+        uint32_t  events;
+        events = k_event_wait_safe(&my_event, 0xFFF, false, K_MSEC(50));
+        if (events == 0) {
+            printk("No input devices are available!");
+        } else {
+            /* Access the desired input device(s) */
+            ...
+        }
+        ...
+    }
+Alternatively, the consumer thread may desire to wait for all the events
+(with atomic removal upon receipt) before continuing using
+:c:func:`k_event_wait_all_safe`.
+
+If all events are received on time, then they will not be present in the event
+object until the next time that the events are set or posted.
+
+.. code-block:: c
+    void consumer_thread(void)
+    {
+        uint32_t  events;
+        events = k_event_wait_all_safe(&my_event, 0x121, false, K_MSEC(50));
+        if (events == 0) {
+            printk("At least one input device is not available!");
+        } else {
+            /* Access the desired input devices */
+            ...
+        }
+        ...
+    }
+
 Suggested Uses
 **************
 
diff --git a/include/zephyr/kernel.h b/include/zephyr/kernel.h
index 858e0a952ba..51140639857 100644
--- a/include/zephyr/kernel.h
+++ b/include/zephyr/kernel.h
@@ -2507,6 +2507,48 @@ __syscall uint32_t k_event_wait(struct k_event *event, uint32_t events,
 __syscall uint32_t k_event_wait_all(struct k_event *event, uint32_t events,
                                    bool reset, k_timeout_t timeout);
 
+/**
+ * @brief Wait for any of the specified events (safe version)
+ *
+ * This call is nearly identical to @ref k_event_wait with the main difference
+ * being that the safe version atomically clears received events from the
+ * event object. This mitigates the need for calling @ref k_event_clear, or
+ * passing a "reset" argument, since doing so may result in lost event
+ * information.
+ *
+ * @param event Address of the event object
+ * @param events Set of desired events on which to wait
+ * @param timeout Waiting period for the desired set of events or one of the
+ *                special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval set of matching events upon success
+ * @retval 0 if no matching event was received within the specified time
+ */
+__syscall uint32_t k_event_wait_safe(struct k_event *event, uint32_t events,
+                                    bool reset, k_timeout_t timeout);
+
+/**
+ * @brief Wait for all of the specified events (safe version)
+ *
+ * This call is nearly identical to @ref k_event_wait_all with the main
+ * difference being that the safe version atomically clears received events
+ * from the event object. This mitigates the need for calling
+ * @ref k_event_clear, or passing a "reset" argument, since doing so may
+ * result in lost event information.
+ *
+ * @param event Address of the event object
+ * @param events Set of desired events on which to wait
+ * @param timeout Waiting period for the desired set of events or one of the
+ *                special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval set of matching events upon success
+ * @retval 0 if all matching events were not received within the specified time
+ */
+__syscall uint32_t k_event_wait_all_safe(struct k_event *event, uint32_t events,
+                                        bool reset, k_timeout_t timeout);
+
+
+
 /**
  * @brief Test the events currently tracked in the event object
  *
diff --git a/include/zephyr/kernel/thread.h b/include/zephyr/kernel/thread.h
index ae43d71115d..a878a283d5c 100644
--- a/include/zephyr/kernel/thread.h
+++ b/include/zephyr/kernel/thread.h
@@ -279,7 +279,8 @@ struct k_thread {
 #if defined(CONFIG_EVENTS)
        struct k_thread *next_event_link;
 
-       uint32_t   events;
+       uint32_t   wevents; /* wait on */
+       uint32_t   revents; /* received */
        uint32_t   event_options;
 
        /** true if timeout should not wake the thread */
diff --git a/kernel/events.c b/kernel/events.c
index 038697d931f..3d5b2951851 100644
--- a/kernel/events.c
+++ b/kernel/events.c
@@ -38,11 +38,13 @@
 #define K_EVENT_WAIT_ALL      0x01   /* Wait for all events */
 #define K_EVENT_WAIT_MASK     0x01
 
-#define K_EVENT_WAIT_RESET    0x02   /* Reset events prior to waiting */
+#define K_EVENT_OPTION_RESET  0x02   /* Reset events prior to waiting */
+#define K_EVENT_OPTION_CLEAR  0x04   /* Clear events that are received */
 
 struct event_walk_data {
        struct k_thread  *head;
        uint32_t events;
+       uint32_t clear_events;
 };
 
 #ifdef CONFIG_OBJ_CORE_EVENT
@@ -77,7 +79,7 @@ void z_vrfy_k_event_init(struct k_event *event)
 #endif /* CONFIG_USERSPACE */
 
 /**
- * @brief determine if desired set of events been satisfied
+ * @brief determine the set of events that have been satisfied
  *
  * This routine determines if the current set of events satisfies the desired
  * set of events. If @a wait_condition is K_EVENT_WAIT_ALL, then at least
@@ -85,31 +87,34 @@ void z_vrfy_k_event_init(struct k_event *event)
  * wait_condition is not K_EVENT_WAIT_ALL, it is assumed to be K_EVENT_WAIT_ANY.
  * In the K_EVENT_WAIT_ANY case, the request is satisfied when any of the
  * current set of events are present in the desired set of events.
+ *
+ * @return event bits that satisfy the wait condition or zero
  */
-static bool are_wait_conditions_met(uint32_t desired, uint32_t current,
-                                   unsigned int wait_condition)
+static uint32_t are_wait_conditions_met(uint32_t desired, uint32_t current,
ckhardin@abyss zephyr % 
ckhardin@abyss zephyr % 
ckhardin@abyss zephyr % 
ckhardin@abyss zephyr % pwd
/Users/ckhardin/git/zephyr
ckhardin@abyss zephyr % less kernel/events.c 
ckhardin@abyss zephyr % less kernel/events.c
ckhardin@abyss zephyr % git log             
commit 8bad1842f6a9b5d0a143b242488b6b92945d4277 (HEAD -> main)
Author: Charles Hardin <ckhardin@gmail.com>
Date:   Fri Aug 29 15:50:49 2025 -0700

    event api that allows events to be cleared and avoids phantom events

commit 0d794df481b9149e1f1a56450f1b3ff610e88477 (upstream/main, origin/main, origin/HEAD)
Author: Sven Ginka <s.ginka@sensry.de>
Date:   Thu Aug 21 10:42:34 2025 +0200

    board: sensry: ganymed-sk - fix missing gpio/NET_L2
    
    when compiling net-samples then the required gpio is disabled.
    the installed vsc8541 phy requires gpio for reset. additionally,
    we enable NET_L2 when network is enabled.
    
    Signed-off-by: Sven Ginka <s.ginka@sensry.de>

commit bf4c8c2e2211c7d93ce668a7662d4536c24d1cc2
Author: Sven Ginka <s.ginka@sensry.de>
Date:   Thu Aug 21 10:42:13 2025 +0200

ckhardin@abyss zephyr % git show --patch
commit 8bad1842f6a9b5d0a143b242488b6b92945d4277 (HEAD -> main)
Author: Charles Hardin <ckhardin@gmail.com>
Date:   Fri Aug 29 15:50:49 2025 -0700

    event api that allows events to be cleared and avoids phantom events

diff --git a/doc/kernel/services/synchronization/events.rst b/doc/kernel/services/synchronization/events.rst
index cbb86894723..c7dc43387ae 100644
--- a/doc/kernel/services/synchronization/events.rst
+++ b/doc/kernel/services/synchronization/events.rst
@@ -108,8 +108,8 @@ the event object.
         ...
     }
 
-Waiting for Events
-==================
+Waiting for Events (without removal)
+====================================
 
 Threads wait for events by calling :c:func:`k_event_wait`.
 
@@ -152,6 +152,53 @@ before continuing.
ckhardin@abyss zephyr % git show --patch | cat
commit 8bad1842f6a9b5d0a143b242488b6b92945d4277
Author: Charles Hardin <ckhardin@gmail.com>
Date:   Fri Aug 29 15:50:49 2025 -0700

    event api that allows events to be cleared and avoids phantom events

diff --git a/doc/kernel/services/synchronization/events.rst b/doc/kernel/services/synchronization/events.rst
index cbb86894723..c7dc43387ae 100644
--- a/doc/kernel/services/synchronization/events.rst
+++ b/doc/kernel/services/synchronization/events.rst
@@ -108,8 +108,8 @@ the event object.
         ...
     }
 
-Waiting for Events
-==================
+Waiting for Events (without removal)
+====================================
 
 Threads wait for events by calling :c:func:`k_event_wait`.
 
@@ -152,6 +152,53 @@ before continuing.
         ...
     }
 
+Waiting for Events (with removal)
+=================================
+
+Threads wait for events (with atomic removal upon receipt) by calling
+:c:func:`k_event_wait_safe`.
+
+The following code builds on the example above, and waits up to 50 milliseconds
+for any of the specified events to be posted.  A warning is issued if none
+of the events are posted in time.
+
+If events are received on time, then they will not be present in the event
+object until the next time that the events are set or posted.
+
+.. code-block:: c
+    void consumer_thread(void)
+    {
+        uint32_t  events;
+        events = k_event_wait_safe(&my_event, 0xFFF, false, K_MSEC(50));
+        if (events == 0) {
+            printk("No input devices are available!");
+        } else {
+            /* Access the desired input device(s) */
+            ...
+        }
+        ...
+    }
+Alternatively, the consumer thread may desire to wait for all the events
+(with atomic removal upon receipt) before continuing using
+:c:func:`k_event_wait_all_safe`.
+
+If all events are received on time, then they will not be present in the event
+object until the next time that the events are set or posted.
+
+.. code-block:: c
+    void consumer_thread(void)
+    {
+        uint32_t  events;
+        events = k_event_wait_all_safe(&my_event, 0x121, false, K_MSEC(50));
+        if (events == 0) {
+            printk("At least one input device is not available!");
+        } else {
+            /* Access the desired input devices */
+            ...
+        }
+        ...
+    }
+
 Suggested Uses
 **************
 
diff --git a/include/zephyr/kernel.h b/include/zephyr/kernel.h
index 858e0a952ba..51140639857 100644
--- a/include/zephyr/kernel.h
+++ b/include/zephyr/kernel.h
@@ -2507,6 +2507,48 @@ __syscall uint32_t k_event_wait(struct k_event *event, uint32_t events,
 __syscall uint32_t k_event_wait_all(struct k_event *event, uint32_t events,
 				    bool reset, k_timeout_t timeout);
 
+/**
+ * @brief Wait for any of the specified events (safe version)
+ *
+ * This call is nearly identical to @ref k_event_wait with the main difference
+ * being that the safe version atomically clears received events from the
+ * event object. This mitigates the need for calling @ref k_event_clear, or
+ * passing a "reset" argument, since doing so may result in lost event
+ * information.
+ *
+ * @param event Address of the event object
+ * @param events Set of desired events on which to wait
+ * @param timeout Waiting period for the desired set of events or one of the
+ *                special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval set of matching events upon success
+ * @retval 0 if no matching event was received within the specified time
+ */
+__syscall uint32_t k_event_wait_safe(struct k_event *event, uint32_t events,
+				     bool reset, k_timeout_t timeout);
+
+/**
+ * @brief Wait for all of the specified events (safe version)
+ *
+ * This call is nearly identical to @ref k_event_wait_all with the main
+ * difference being that the safe version atomically clears received events
+ * from the event object. This mitigates the need for calling
+ * @ref k_event_clear, or passing a "reset" argument, since doing so may
+ * result in lost event information.
+ *
+ * @param event Address of the event object
+ * @param events Set of desired events on which to wait
+ * @param timeout Waiting period for the desired set of events or one of the
+ *                special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval set of matching events upon success
+ * @retval 0 if all matching events were not received within the specified time
+ */
+__syscall uint32_t k_event_wait_all_safe(struct k_event *event, uint32_t events,
+					 bool reset, k_timeout_t timeout);
+
+
+
 /**
  * @brief Test the events currently tracked in the event object
  *
diff --git a/include/zephyr/kernel/thread.h b/include/zephyr/kernel/thread.h
index ae43d71115d..a878a283d5c 100644
--- a/include/zephyr/kernel/thread.h
+++ b/include/zephyr/kernel/thread.h
@@ -279,7 +279,8 @@ struct k_thread {
 #if defined(CONFIG_EVENTS)
 	struct k_thread *next_event_link;
 
-	uint32_t   events;
+	uint32_t   wevents; /* wait on */
+	uint32_t   revents; /* received */
 	uint32_t   event_options;
 
 	/** true if timeout should not wake the thread */
diff --git a/kernel/events.c b/kernel/events.c
index 038697d931f..3d5b2951851 100644
--- a/kernel/events.c
+++ b/kernel/events.c
@@ -38,11 +38,13 @@
 #define K_EVENT_WAIT_ALL      0x01   /* Wait for all events */
 #define K_EVENT_WAIT_MASK     0x01
 
-#define K_EVENT_WAIT_RESET    0x02   /* Reset events prior to waiting */
+#define K_EVENT_OPTION_RESET  0x02   /* Reset events prior to waiting */
+#define K_EVENT_OPTION_CLEAR  0x04   /* Clear events that are received */
 
 struct event_walk_data {
 	struct k_thread  *head;
 	uint32_t events;
+	uint32_t clear_events;
 };
 
 #ifdef CONFIG_OBJ_CORE_EVENT
@@ -77,7 +79,7 @@ void z_vrfy_k_event_init(struct k_event *event)
 #endif /* CONFIG_USERSPACE */
 
 /**
- * @brief determine if desired set of events been satisfied
+ * @brief determine the set of events that have been satisfied
  *
  * This routine determines if the current set of events satisfies the desired
  * set of events. If @a wait_condition is K_EVENT_WAIT_ALL, then at least
@@ -85,31 +87,34 @@ void z_vrfy_k_event_init(struct k_event *event)
  * wait_condition is not K_EVENT_WAIT_ALL, it is assumed to be K_EVENT_WAIT_ANY.
  * In the K_EVENT_WAIT_ANY case, the request is satisfied when any of the
  * current set of events are present in the desired set of events.
+ *
+ * @return event bits that satisfy the wait condition or zero
  */
-static bool are_wait_conditions_met(uint32_t desired, uint32_t current,
-				    unsigned int wait_condition)
+static uint32_t are_wait_conditions_met(uint32_t desired, uint32_t current,
+					unsigned int wait_condition)
 {
-	uint32_t  match = current & desired;
+	uint32_t match = current & desired;
 
-	if (wait_condition == K_EVENT_WAIT_ALL) {
-		return match == desired;
+	if ((wait_condition == K_EVENT_WAIT_ALL) && (match != desired)) {
+		/* special case for K_EVENT_WAIT_ALL */
+		return 0;
 	}
 
-	/* wait_condition assumed to be K_EVENT_WAIT_ANY */
-
-	return match != 0;
+	/* return the matched events for any wait condition */
+	return match;
 }
 
 static int event_walk_op(struct k_thread *thread, void *data)
 {
-	unsigned int      wait_condition;
+	uint32_t match;
+	unsigned int wait_condition;
 	struct event_walk_data *event_data = data;
 
 	wait_condition = thread->event_options & K_EVENT_WAIT_MASK;
 
-	if (are_wait_conditions_met(thread->events, event_data->events,
-				    wait_condition)) {
-
+	match = are_wait_conditions_met(thread->wevents, event_data->events,
+					wait_condition);
+	if (match != 0) {
 		/*
 		 * Events create a list of threads to wake up. We do
 		 * not want z_thread_timeout to wake these threads; they
@@ -117,14 +122,21 @@ static int event_walk_op(struct k_thread *thread, void *data)
 		 * have been processed.
 		 */
 		thread->no_wake_on_timeout = true;
+		z_abort_timeout(&thread->base.timeout);
 
 		/*
-		 * The wait conditions have been satisfied. Add this
-		 * thread to the list of threads to unpend.
+		 * The wait conditions have been satisfied. So, set the
+		 * received events and then add this thread to the list
+		 * of threads to unpend.
+		 *
+		 * NOTE: thread event options can consume an event
 		 */
+		thread->revents = match;
+		if (thread->event_options & K_EVENT_OPTION_CLEAR) {
+			event_data->clear_events |= match;
+		}
 		thread->next_event_link = event_data->head;
 		event_data->head = thread;
-		z_abort_timeout(&thread->base.timeout);
 	}
 
 	return 0;
@@ -147,8 +159,7 @@ static uint32_t k_event_post_internal(struct k_event *event, uint32_t events,
 	previous_events = event->events & events_mask;
 	events = (event->events & ~events_mask) |
 		 (events & events_mask);
-	event->events = events;
-	data.events = events;
+
 	/*
 	 * Posting an event has the potential to wake multiple pended threads.
 	 * It is desirable to unpend all affected threads simultaneously. This
@@ -159,6 +170,8 @@ static uint32_t k_event_post_internal(struct k_event *event, uint32_t events,
 	 * 3. Ready each of the threads in the linked list
 	 */
 
+	data.events = events;
+	data.clear_events = 0;
 	z_sched_waitq_walk(&event->wait_q, event_walk_op, &data);
 
 	if (data.head != NULL) {
@@ -166,13 +179,15 @@ static uint32_t k_event_post_internal(struct k_event *event, uint32_t events,
 		struct k_thread *next;
 		do {
 			arch_thread_return_value_set(thread, 0);
-			thread->events = events;
 			next = thread->next_event_link;
 			z_sched_wake_thread(thread, false);
 			thread = next;
 		} while (thread != NULL);
 	}
 
+	/* stash any events not consumed */
+	event->events = data.events & ~data.clear_events;
+
 	z_reschedule(&event->lock, key);
 
 	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_event, post, event, events,
@@ -262,21 +277,22 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
 
 	k_spinlock_key_t  key = k_spin_lock(&event->lock);
 
-	if (options & K_EVENT_WAIT_RESET) {
+	if (options & K_EVENT_OPTION_RESET) {
 		event->events = 0;
 	}
 
 	/* Test if the wait conditions have already been met. */
-
-	if (are_wait_conditions_met(events, event->events, wait_condition)) {
-		rv = event->events;
+	rv = are_wait_conditions_met(events, event->events, wait_condition);
+	if (rv != 0) {
+		/* clear the events that are matched */
+		if (options & K_EVENT_OPTION_CLEAR) {
+			event->events &= ~rv;
+		}
 
 		k_spin_unlock(&event->lock, key);
 		goto out;
 	}
 
-	/* Match conditions have not been met. */
-
 	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
 		k_spin_unlock(&event->lock, key);
 		goto out;
@@ -287,7 +303,8 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
 	 * set of events in the k_thread structure.
 	 */
 
-	thread->events = events;
+	thread->wevents = events;
+	thread->revents = 0;
 	thread->event_options = options;
 
 	SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_event, wait, event, events,
@@ -295,14 +312,13 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
 
 	if (z_pend_curr(&event->lock, key, &event->wait_q, timeout) == 0) {
 		/* Retrieve the set of events that woke the thread */
-		rv = thread->events;
+		rv = thread->revents;
 	}
 
 out:
-	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_event, wait, event,
-				       events, rv & events);
+	SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_event, wait, event, events, rv);
 
-	return rv & events;
+	return rv;
 }
 
 /**
@@ -311,7 +327,7 @@ out:
 uint32_t z_impl_k_event_wait(struct k_event *event, uint32_t events,
 			     bool reset, k_timeout_t timeout)
 {
-	uint32_t options = reset ? K_EVENT_WAIT_RESET : 0;
+	uint32_t options = reset ? K_EVENT_OPTION_RESET : 0;
 
 	return k_event_wait_internal(event, events, options, timeout);
 }
@@ -331,7 +347,7 @@ uint32_t z_vrfy_k_event_wait(struct k_event *event, uint32_t events,
 uint32_t z_impl_k_event_wait_all(struct k_event *event, uint32_t events,
 				 bool reset, k_timeout_t timeout)
 {
-	uint32_t options = reset ? (K_EVENT_WAIT_RESET | K_EVENT_WAIT_ALL)
+	uint32_t options = reset ? (K_EVENT_OPTION_RESET | K_EVENT_WAIT_ALL)
 				 : K_EVENT_WAIT_ALL;
 
 	return k_event_wait_internal(event, events, options, timeout);
@@ -347,6 +363,45 @@ uint32_t z_vrfy_k_event_wait_all(struct k_event *event, uint32_t events,
 #include <zephyr/syscalls/k_event_wait_all_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
+uint32_t z_impl_k_event_wait_safe(struct k_event *event, uint32_t events,
+				  bool reset, k_timeout_t timeout)
+{
+	uint32_t options = reset ? (K_EVENT_OPTION_CLEAR | K_EVENT_OPTION_RESET)
+				 : K_EVENT_OPTION_CLEAR;
+
+	return k_event_wait_internal(event, events, options, timeout);
+}
+
+#ifdef CONFIG_USERSPACE
+uint32_t z_vrfy_k_event_wait_safe(struct k_event *event, uint32_t events,
+				  bool reset, k_timeout_t timeout)
+{
+	K_OOPS(K_SYSCALL_OBJ(event, K_OBJ_EVENT));
+	return z_impl_k_event_wait_safe(event, events, reset, timeout);
+}
+#include <zephyr/syscalls/k_event_wait_safe_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+uint32_t z_impl_k_event_wait_all_safe(struct k_event *event, uint32_t events,
+				      bool reset, k_timeout_t timeout)
+{
+	uint32_t options = reset ? (K_EVENT_OPTION_CLEAR |
+				    K_EVENT_OPTION_RESET | K_EVENT_WAIT_ALL)
+				 : (K_EVENT_OPTION_CLEAR | K_EVENT_WAIT_ALL);
+
+	return k_event_wait_internal(event, events, options, timeout);
+}
+
+#ifdef CONFIG_USERSPACE
+uint32_t z_vrfy_k_event_wait_all_safe(struct k_event *event, uint32_t events,
+				      bool reset, k_timeout_t timeout)
+{
+	K_OOPS(K_SYSCALL_OBJ(event, K_OBJ_EVENT));
+	return z_impl_k_event_wait_all_safe(event, events, reset, timeout);
+}
+#include <zephyr/syscalls/k_event_wait_all_safe_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
 #ifdef CONFIG_OBJ_CORE_EVENT
 static int init_event_obj_core_list(void)
 {
diff --git a/tests/kernel/events/event_api/src/test_event_apis.c b/tests/kernel/events/event_api/src/test_event_apis.c
index 818f254f40d..5a4b19c60b1 100644
--- a/tests/kernel/events/event_api/src/test_event_apis.c
+++ b/tests/kernel/events/event_api/src/test_event_apis.c
@@ -434,6 +434,45 @@ ZTEST(events_api, test_event_receive)
 
 	test_wake_multiple_threads();
 }
+
+TEST(events_api, test_k_event_wait_safe)
+{
+	uint32_t events;
+
+	k_event_set(&test_event, 0x73);
+
+	events = k_event_wait_safe(&test_event, 0x11, false, K_NO_WAIT);
+	zexpect_equal(events, 0x11, "expected 0x11, got %x", events);
+
+	events = k_event_wait_safe(&test_event, 0x11, false, K_NO_WAIT);
+	zexpect_equal(events, 0x0, "phantom events %x not removed from event object", events);
+
+	events = k_event_wait_safe(&test_event, 0x62, false, K_NO_WAIT);
+	zexpect_equal(events, 0x62, "expected 0x62, got %x", events);
+
+	events = k_event_wait_safe(&test_event, -1, false, K_NO_WAIT);
+	zexpect_equal(events, 0x0, "phantom events %x not removed from event object", events);
+}
+
+ZTEST(events_api, test_k_event_wait_all_safe)
+{
+	uint32_t events;
+
+	k_event_set(&test_event, 0x73);
+
+	events = k_event_wait_all_safe(&test_event, 0x81, false, K_NO_WAIT);
+	zexpect_equal(events, 0x0, "expected 0x0, got %x", events);
+
+	events = k_event_wait_all_safe(&test_event, 0x11, false, K_NO_WAIT);
+	zexpect_equal(events, 0x11, "expected 0x11, got %x", events);
+
+	events = k_event_wait_all_safe(&test_event, 0x63, false, K_NO_WAIT);
+	zexpect_equal(events, 0x0, "expected 0x0, got %x", events);
+
+	events = k_event_wait_all_safe(&test_event, 0x62, false, K_NO_WAIT);
+	zexpect_equal(events, 0x62, "expected 0x62, got %x", events);
+}
+
 /**
  * @}
  */
```
